### PR TITLE
add fragment to init profile photo+function photo in userModel+functi…

### LINKED ISSUE #59: Add the possibility to the user to add a profile photo to his user profile.

### PR DESCRIPTION
**Pull request by HugoHof**
This PR add to the user the possibility to add a profile pho in his user account. So this extension contains:
- a fragment ProfilePhotoInitFrag to potentially add a profile photo to a user account. 
- modify the constructor and UserModel to potentially add an optional profile photo.
- function in database to set a profile photo to a user account.

Without image selection:
Capture d’écran 2023-03-28 à 17 03 21

With image preview: 
Capture d’écran 2023-03-28 à 17 04 18

During image selecting : 
Capture d’écran 2023-03-28 à 17 04 32

Thank you for your review

Remark: the image are store in the firebase as a string(encoding in base64)
I don't arrive to test the image selector on the fragment since we must have some image present on the local memory. I have try to create custom image library but I have problem with write permission on the phone disk(which will cause problem on CI). So I haven't tested directly the image picker.
close #59

### CODE REVIEW

File: app/src/main/java/com/epfl/drawyourpath/database/Database.kt [hunk 1]
```diff
@@ -1,6 +1,8 @@
 package com.epfl.drawyourpath.database
 
 
+import android.graphics.Bitmap
+import android.provider.ContactsContract.Contacts.Photo
```
Comment (HoLuDyEr) [inline app/src/main/java/com/epfl/drawyourpath/database/Database.kt:5]: detail: unused imports

File: app/src/main/java/com/epfl/drawyourpath/database/Database.kt [hunk 2]
```diff
@@ -1,6 +1,8 @@
...
 import com.epfl.drawyourpath.userProfile.UserModel
 import java.time.LocalDate
 import java.util.*
```
Comment (HoLuDyEr) [inline app/src/main/java/com/epfl/drawyourpath/database/Database.kt:8]: detail: unused imports

File: app/src/main/java/com/epfl/drawyourpath/database/FireDatabase.kt [hunk 3]
```diff
@@ -1,5 +1,7 @@
 package com.epfl.drawyourpath.database
 
+import android.graphics.Bitmap
+import android.graphics.BitmapFactory
 import com.epfl.drawyourpath.authentication.Auth
```
Comment (HoLuDyEr) [inline app/src/main/java/com/epfl/drawyourpath/database/FireDatabase.kt:5]: detail: unused imports

File: app/src/main/java/com/epfl/drawyourpath/database/FireDatabase.kt [hunk 4]
```diff
@@ -8,7 +10,9 @@ import com.google.firebase.database.DataSnapshot
 import com.google.firebase.database.DatabaseReference
 import com.google.firebase.database.ktx.database
 import com.google.firebase.ktx.Firebase
+import java.io.ByteArrayOutputStream
 import java.time.LocalDate
+import java.util.Base64
 import java.util.concurrent.CompletableFuture
 
 private val TIMEOUT_SERVER_REQUEST: Long = 10
```
Comment (HoLuDyEr) [inline app/src/main/java/com/epfl/drawyourpath/database/FireDatabase.kt:18]: this variable is never used, could it be deleted?

File: app/src/main/java/com/epfl/drawyourpath/database/MockDataBase.kt [hunk 5]
```diff
@@ -1,5 +1,7 @@
 package com.epfl.drawyourpath.database
 
+import android.graphics.Bitmap
+import android.graphics.BitmapFactory
```
Comment (HoLuDyEr) [inline app/src/main/java/com/epfl/drawyourpath/database/MockDataBase.kt:4]: detail: unused imports

File: app/src/main/java/com/epfl/drawyourpath/userProfile/UserModel.kt [hunk 6]
```diff
@@ -313,6 +318,28 @@ class UserModel {
...
+    /**
+     * This function will  return the profile photo of the user or null if it doesn't exit
+     * @return the profile photo if it exist
+     */
+    fun getProfilePhoto(): Bitmap?{
+        return profilePhoto
+    }
+
+    /**
+     * This function will  the photo as the profile photo of the user
```
Comment (HoLuDyEr) [inline app/src/main/java/com/epfl/drawyourpath/userProfile/UserModel.kt:330]: detail: typo
This function will **_set_** the photo

File: app/src/test/java/com/epfl/drawyourpath/userProfile/UserModelTest.kt [hunk 7]
```diff
@@ -1,5 +1,11 @@
 package com.epfl.drawyourpath.bootcamp.userProfile
```
Comment (HoLuDyEr) [inline app/src/test/java/com/epfl/drawyourpath/userProfile/UserModelTest.kt:1]: wrong package name

File: app/src/test/java/com/epfl/drawyourpath/userProfile/UserModelTest.kt [hunk 8]
```diff
@@ -1,5 +1,11 @@
...
 
+import android.graphics.Bitmap
+import android.graphics.BitmapFactory
+import android.graphics.Color
+import android.graphics.Picture
+import android.media.Image
+import androidx.core.graphics.set
```
Comment (HoLuDyEr) [inline app/src/test/java/com/epfl/drawyourpath/userProfile/UserModelTest.kt:8]: detail: unused imports

File: app/src/main/java/com/epfl/drawyourpath/userProfileCreation/PhotoProfileInitFragment.kt [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+        selectPhotoButton.setOnClickListener{
+            val photoPicker = Intent(Intent.ACTION_PICK, MediaStore.Images.Media.INTERNAL_CONTENT_URI)
+            startActivityForResult(photoPicker, requestCodeFrag)
+        }
+
+        //if we click on the skip button we simply pass the end profile creation fragment
+        val skipButton: Button = view.findViewById(R.id.skipPhotoProfile_button_userProfileCreation)
+        skipButton.setOnClickListener{
+            showEndProfileCreationFrag(database)
+        }
```
Comment (HoLuDyEr) [inline app/src/main/java/com/epfl/drawyourpath/userProfileCreation/PhotoProfileInitFragment.kt:66]: could be moved to a private function
`private fun createSkipButton(view: View, database: Database)`

File: app/src/main/java/com/epfl/drawyourpath/userProfileCreation/PhotoProfileInitFragment.kt [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+     * Helper function to show a message if the user forgot to select a photo and store the photo in the database if the photo is not null
+     * @param database used to store the photo
+     * @return a future to indicate if the photo was store in the database
+     */
+    private fun isPhotoSelected(database: Database): CompletableFuture<Boolean>{
+        val future = CompletableFuture<Boolean>()
+
+        if(photoProfile==null){
+
+            errorText.text = "* You have forget to select a photo !"
```
Comment (HoLuDyEr) [inline app/src/main/java/com/epfl/drawyourpath/userProfileCreation/PhotoProfileInitFragment.kt:103]: detail: typo
You have **_forgotten_** to select a photo

File: app/src/main/java/com/epfl/drawyourpath/userProfileCreation/PhotoProfileInitFragment.kt [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+import android.os.Bundle
+import android.provider.MediaStore
+import android.view.View
+import android.widget.Button
+import android.widget.ImageView
+import android.widget.TextView
+import androidx.fragment.app.Fragment
+import com.epfl.drawyourpath.R
+import com.epfl.drawyourpath.authentication.FirebaseAuth
+import com.epfl.drawyourpath.authentication.MockAuth
```
Comment (HoLuDyEr) [inline app/src/main/java/com/epfl/drawyourpath/userProfileCreation/PhotoProfileInitFragment.kt:17]: detail: unused imports

File: app/src/main/java/com/epfl/drawyourpath/userProfileCreation/PhotoProfileInitFragment.kt [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+import com.epfl.drawyourpath.database.Database
+import com.epfl.drawyourpath.database.FireDatabase
+import com.epfl.drawyourpath.database.MockDataBase
+import com.epfl.drawyourpath.login.LoginActivity
```
Comment (HoLuDyEr) [inline app/src/main/java/com/epfl/drawyourpath/userProfileCreation/PhotoProfileInitFragment.kt:21]: detail: unused imports

**Review by HoLuDyEr — CHANGES_REQUESTED** [review]
Overall it is a good PR, the code is clean with good documentation.
There are a few details to fix.

I would suggest that you try to use Code -> Reformat Code (crtl + alt+ L / cmd + alt + L) or Code -> Optimize Import (crtl + alt + O / cmd + alt + O)

**Comment by HugoHof** [comment]
> Overall it is a good PR, the code is clean with good documentation. There are a few details to fix.
> 
> I would suggest that you try to use Code -> Reformat Code (crtl + alt+ L / cmd + alt + L) or Code -> Optimize Import (crtl + alt + O / cmd + alt + O)

Thank you for your review. Sorry, I have forgotten to refactor the code. I have modify my code with all your advice.

**Review by HoLuDyEr — APPROVED** [review]
Thank you for the changes, now it is ready to be merged

**Comment by HugoHof** [comment]
> I just have a note concerning how we save the images. Images shouldn't be put in the Realtime Database. We can download 360MB daily from the database and adding even just a 1MB image in a user profile will make us reach this threshold really fast. The recommended way to store image profiles is to use [Firebase Storage](https://firebase.google.com/docs/storage).
> 
> We might add the conversion for a task in the next sprint. What do you think?

Thank you for your idea. We can discuss it on Friday. I haven't think about using the firebase storage. It's a good idea.

**Pull request merged**